src/library/dao/macrodao.cpp

```diff
@@ -0,0 +1,94 @@
+#include "library/dao/macrodao.h"
+
+#include <QDebug>
+#include <QSqlQuery>
+
+#include "library/queryutil.h"
+
+bool MacroDAO::saveMacro(TrackId trackId, Macro* macro, int slot) const {
+    QSqlQuery query(m_database);
+    DEBUG_ASSERT(slot > 0);
+    if (macro->getId() == DbId::s_invalidValue) {
+        query.prepare(QStringLiteral(
+                "INSERT INTO macros "
+                "(track_id, slot, label, state, content) "
+                "VALUES "
+                "(:track_id, :slot, :label, :state, :content)"));
+    } else {
+        query.prepare(QStringLiteral(
+                "UPDATE macros SET "
+                "track_id=:track_id,"
+                "slot=:slot,"
+                "label=:label,"
+                "state=:state,"
+                "content=:content"
+                " WHERE id=:id"));
+        query.bindValue(":id", macro->getId());
+    }
+    query.bindValue(":track_id", trackId.toVariant());
+    query.bindValue(":slot", slot);
+    query.bindValue(":label", macro->getLabel());
+    query.bindValue(":state", macro->getState().operator int());
+    query.bindValue(":content", Macro::serialize(macro->getActions()));
+
+    if (query.exec()) {
+        if (macro->getId() == DbId::s_invalidValue) {
+            macro->setId(query.lastInsertId().toInt());
+        }
+        macro->setDirty(false);
+        return true;
+    } else {
+        LOG_FAILED_QUERY(query);
+        return false;
+    }
+}
+
+void MacroDAO::saveMacros(TrackId trackId, const QMap<int, MacroPointer>& macros) const {
+    for (auto it = macros.constBegin(); it != macros.constEnd(); ++it) {
+        auto pMacro = it.value();
+        // Don't save placeholder Macros
+        if (pMacro->isEmpty() && !pMacro->isDirty()) {
+            return;
+        }
+        // Newly recorded macros must be dirty
+        DEBUG_ASSERT(pMacro->getId() >= 0 || pMacro->isDirty());
+        if (pMacro->isDirty()) {
+            saveMacro(trackId, pMacro.get(), it.key());
+        }
+        // After saving each macro must have a valid id and not be dirty
+        DEBUG_ASSERT(pMacro->getId() >= 0 && !pMacro->isDirty());
+    }
+}
+
+QSqlQuery MacroDAO::querySelect(QString columns, TrackId trackId) const {
```

Comment:
Clazy complains about this, you need to pass QString by const reference

Reply:
I have run pre-commit, if that doesn't do that it should be added

Reply:
clazy is terribly slow. We can't add it to the pre-commit hooks unless you'd like to wait 2 hours between each commit.